core/shared/src/main/scala/fs2/Stream.scala

```diff
@@ -2134,43 +2141,59 @@ object Stream {
       * }}}
       */
     def merge[O2 >: O](that: Stream[F, O2])(implicit F: Effect[F],
-                                            ec: ExecutionContext): Stream[F, O2] = {
-      def go(l: AsyncPull[F, Option[(Segment[O2, Unit], Stream[F, O2])]],
-             r: AsyncPull[F, Option[(Segment[O2, Unit], Stream[F, O2])]]): Pull[F, O2, Unit] =
-        l.race(r).pull.flatMap {
-          case Left(l) =>
-            l match {
-              case None =>
-                r.pull.flatMap {
-                  case None           => Pull.done
-                  case Some((hd, tl)) => Pull.output(hd) >> tl.pull.echo
-                }
-              case Some((hd, tl)) =>
-                Pull.output(hd) >> tl.pull.unconsAsync.flatMap(go(_, r))
-            }
-          case Right(r) =>
-            r match {
-              case None =>
-                l.pull.flatMap {
-                  case None           => Pull.done
-                  case Some((hd, tl)) => Pull.output(hd) >> tl.pull.echo
+                                            ec: ExecutionContext): Stream[F, O2] =
+      Stream.eval(async.semaphore(0)).flatMap { doneSem =>
+        Stream.eval(async.promise[F, Unit]).flatMap { interruptL =>
+          Stream.eval(async.promise[F, Unit]).flatMap { interruptR =>
+            Stream.eval(async.promise[F, Throwable]).flatMap { interruptY =>
+              Stream
+                .eval(async.unboundedQueue[F, Option[(F[Unit], Chunk[O2])]])
+                .flatMap { outQ => // note that the queue actually contains up to 2 max elements thanks to semaphores guarding each side.
+                  def runUpstream(s: Stream[F, O2], interrupt: Promise[F, Unit]): F[Unit] =
+                    async.semaphore(1).flatMap { quard =>
+                      s.chunks
+                        .interruptWhen(interrupt.get.attempt)
+                        .evalMap { chunk =>
+                          if (chunk.isEmpty) F.pure(())
```

Comment:
Why elide empty chunks?

Reply:
If we don't do it, stream may in fact hang with exception unable to propagate it to downstream. I had new merge (that I have added) tests failing w/o this, as some of the `Failed` Stream examples seems to emit empty chunk and then failure. I don't think so this does any harm, because if you care about empty chunks, you could always do s1.chunk.merge(s2.chunk) and they won't be elided in that case. For the Stream[F,A] I think use doesn't care about empty chunks?

Reply:
Hm, this makes me thing there's something subtly wrong. We don't elide empty chunks and we use segments in `join` -- why would `merge` be different?

Reply:
ok, I'll investigate this again and see exact cause of this.

Reply:
Ok I remmebered the scenario why we have to emit empty streams. Essentially this comes from the following spec :
```scala
(s1: PureStream[Int], f: Failure) =>
        an[Err.type] should be thrownBy {
          s1.get
            .merge(f.get)
            .flatMap { _ =>
              Stream.eval(IO.async[Unit](_ => ()))
            }
            .compile
            .drain
            .unsafeRunSync()
        }

``` 
if `f` stream is structure of `Stream.emits() ++ Stream.fail` or similar, the stream will not fail unless that empty chunk get processed. That may not be a big deal for that example code, but Empty chunks may not be that explicit as you see there, they in fact may be a result of stream like ` Stream.emit(42).map(_ => throw Err))` where the exception appeared, was thrown but until the empty chunk preceding the failure processing(`Algebra.raiseError())is processed, stream won't fail. 

I think the `join` may perhaps need similar treatment, hence I didn't really wrote tests of its behaviour on blocking `flatMap`.

Reply:
I'm not convinced. I think both `merge` and `join` should operate on segment structure and emit empty segments. As a property, this looks something like `s1.either(s2).segments.compile.toVector.separate == (s1.segments.compile.toVector, s2.segments.compile.toVector)`

Reply:
@mpilquist generally no problem with that, however we have then to release some properties, as the f: Failure currently prohibits certain tests to terminate. I mean I can always try to do `f.get.drain` instead of only `f.get` and that shall likely to work. Any thoughts?

Reply:
Could you do `f.get.chunks.filter(_.nonEmpty)` in the impacted tests?

Reply:
sure, yes, no problem with that although, still, b/c if you look on Failure examples do you think user will have an idea there are empty chunks emitted before failure ? I mean how he should know in case of deadlocks that this `filter` has to be added for the combinator to work as expected? Again just reassuring myself that your suggested path is correct.

Reply:
There's nothing particularly special about empty chunks leading to deadlocks due to the shape of the stream after merge. You could construct an example that deadlocks on singleton chunks or any other shape.

Reply:
ok well, lets start with this and well see how that goes. In fact, the effects shall be run by concurrently anyhow where that should not be an issue at all.

Reply:
@mpilquist seems that only `sink` tests are affected when I change this. So, I gonna revert it back w/o emptiness guard (I still leave this as chunk for (b) reason mentioned above). ok?

Reply:
I think it should be at segment level -- you can always resegment/rechunk to get more concurrency out of it if workload calls for it.